### PR TITLE
Implemented super call rule (#803)

### PR DESCRIPTION
**Pull request by angelolloqui**
PR for issue #803 

It basically looks for methods in the whitelist (and marked as `overriden`) and checks that there is exactly one `super` call somewhere in its implementation.

The rule comes with a custom provider that allows configuring the list of methods to include or exclude. The default configuration comes with the most common methods in Cocoa.

Example only checking 2 custom methods:
```
overridden_super_call:
  severity: warning
  excluded: *
  included: 
    - testMethod1()
    - testMethod2(_:)
```

### CODE REVIEW

**Comment by angelolloqui** [comment]
Current predefined list of included methods:

```
var defaultIncluded = [
        //NSObject
        "awakeFromNib()",
        "prepareForInterfaceBuilder()",
        //UIViewController
        "transitionCoordinator()",
        "viewDidLoad()",
        "viewWillAppear(_:)",
        "viewWillDisappear(_:)",
        "viewDidAppear(_:)",
        "viewDidDisappear(_:)",
        "addChildViewController(_:)",
        "removeFromParentViewController()",
        "transition(from:to:duration:options:animations:completion:)",
        "transitionFromViewController(_:toViewController:duration:options:animations:completion:)",
        "didReceiveMemoryWarning()",
        "encodeRestorableState(with:)",
        "encodeRestorableStateWithCoder(_:)",
        "decodeRestorableState(with:)",
        "decodeRestorableStateWithCoder(_:)",
        "setEditing(_:animated:)",
        //UIView
        "updateConstraints()",
        "prepareForReuse()",
        //UICollectionViewLayout
        "invalidateLayout()",
        "invalidateLayout(with:)",
        "invalidateLayoutWithContext(_:)",
        //XCTestCase
        "tearDown()",
        "setUp()"
    ]
```

File: Source/SwiftLintFramework/Rules/SuperCallRule.swift [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            "}\n"
+        ],
+        triggeringExamples: [
+            "class VC: UIViewController {\n" +
+                "\toverride func viewWillAppear(animated: Bool) () ↓{\n" +
+                    "\t\t//Not calling to super\n" +
+                    "\t\tself.method()\n" +
+                "\t}\n" +
+            "}\n",
+        ]
```
Comment (norio-nomura) [inline Source/SwiftLintFramework/Rules/SuperCallRule.swift]: Could you please add multiple calling to super triggering example?

File: Source/SwiftLintFramework/Rules/RuleConfigurations/SuperCallConfiguration.swift [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        //UIView
+        "updateConstraints()",
+        "prepareForReuse()",
+        //UICollectionViewLayout
+        "invalidateLayout()",
+        "invalidateLayout(with:)",
+        "invalidateLayoutWithContext(_:)",
+        //XCTestCase
+        "tearDown()",
+        "setUp()"
```
Comment (marcelofabri) [inline Source/SwiftLintFramework/Rules/RuleConfigurations/SuperCallConfiguration.swift]: I haven't seen anywhere explicitly, but the default template includes calling super: https://developer.apple.com/library/content/documentation/DeveloperTools/Conceptual/testing_with_xcode/chapters/04-writing_tests.html#//apple_ref/doc/uid/TP40014132-CH4-SW2

File: Source/SwiftLintFramework/Rules/SuperCallRule.swift [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    )
+
+    public func validateFile(file: File,
+                             kind: SwiftDeclarationKind,
+                             dictionary: [String: SourceKitRepresentable]) -> [StyleViolation] {
+
+        guard   let offset = dictionary["key.bodyoffset"] as? Int64,
+                let name = dictionary["key.name"] as? String,
+                let substructure = dictionary["key.substructure"] as? [SourceKitRepresentable]
```
Comment (norio-nomura) [inline Source/SwiftLintFramework/Rules/SuperCallRule.swift]: I found false negative pattern:
```swift
class ViewController: UIViewController {
    override func didReceiveMemoryWarning() {
    }
}
```
that does not have `"key.substructure"`.

Comment (angelolloqui) [inline Source/SwiftLintFramework/Rules/SuperCallRule.swift]: ouch! thanks for pointing this out! I will check it out ASAP

File: Source/SwiftLintFramework/Rules/RuleConfigurations/SuperCallConfiguration.swift [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        if let included = [String].arrayOf(configuration["included"]) {
+            self.included = included
+        }
+    }
+
+    public var severity: ViolationSeverity {
+        return severityConfiguration.severity
+    }
+
+    public var resolvedMethodNames: [String] {
```
Comment (angelolloqui) [inline Source/SwiftLintFramework/Rules/RuleConfigurations/SuperCallConfiguration.swift]: 👍

**Comment by angelolloqui** [comment]
Take a look now... 1 thing I am not sure if should be changed is that the rule id is `overridden_super_call` but the class implementing is `SuperCallRule`. If you want consistency, which name do you think makes more sense? maybe the first one?

**Comment by norio-nomura** [comment]
> maybe the first one?

👍

**Comment by angelolloqui** [comment]
Renamed! Let's hope this now is good enough :)

**Comment by norio-nomura** [comment]
Thanks! 🙏

**Pull request merged**